src/main/java/com/maxmind/db/MultiBuffer.java

```diff
@@ -0,0 +1,364 @@
+        }
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte get() {
+        byte value = get(position);
+        position++;
+        return value;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Buffer get(byte[] dst) {
+        if (position + dst.length > limit) {
+            throw new IndexOutOfBoundsException(
+                    "Read exceeds limit: position=" + position
+                            + ", length=" + dst.length
+                            + ", limit=" + limit
+            );
+        }
+        long pos = position;
+        int offset = 0;
+        int length = dst.length;
+        while (length > 0) {
+            int bufIndex = (int) (pos / this.chunkSize);
+            int bufOffset = (int) (pos % this.chunkSize);
+            ByteBuffer buf = buffers[bufIndex];
+            buf.position(bufOffset);
+            int toRead = Math.min(buf.remaining(), length);
+            buf.get(dst, offset, toRead);
+            pos += toRead;
+            offset += toRead;
+            length -= toRead;
+        }
+        position = pos;
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte get(long index) {
+        if (index < 0 || index >= limit) {
+            throw new IndexOutOfBoundsException("Index: " + index);
+        }
+        int bufIndex = (int) (index / this.chunkSize);
+        int offset = (int) (index % this.chunkSize);
+        return buffers[bufIndex].get(offset);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public double getDouble() {
+        int bufIndex = (int) (position / this.chunkSize);
+        int off = (int) (position % this.chunkSize);
+        ByteBuffer buf = buffers[bufIndex];
+        buf.position(off);
+        if (buf.remaining() >= 8) {
+            double value = buf.getDouble();
+            position += 8;
+            return value;
+        } else {
+            byte[] eight = new byte[8];
+            get(eight);
+            return ByteBuffer.wrap(eight).getDouble();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public float getFloat() {
+        int bufIndex = (int) (position / this.chunkSize);
+        int off = (int) (position % this.chunkSize);
+        ByteBuffer buf = buffers[bufIndex];
+        buf.position(off);
+        if (buf.remaining() >= 4) {
+            float value = buf.getFloat();
+            position += 4;
+            return value;
+        } else {
+            byte[] four = new byte[4];
+            get(four);
+            return ByteBuffer.wrap(four).getFloat();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Buffer duplicate() {
+        MultiBuffer copy = new MultiBuffer(capacity, chunkSize);
+        for (int i = 0; i < buffers.length; i++) {
+            copy.buffers[i] = buffers[i].duplicate();
+        }
+        copy.position = this.position;
+        copy.limit = this.limit;
+        return copy;
```

Comment:
Won't `new MultiBuffer(capacity, chunkSize)` cause us to allocate large buffers? Maybe the JVM is smart enough to optimize this away if the buffers are not modified, but I was expecting something more like this:


```suggestion
        ByteBuffer[] duplicatedBuffers = new ByteBuffer[buffers.length];
        for (int i = 0; i < buffers.length; i++) {
            duplicatedBuffers[i] = buffers[i].duplicate()
        }
        MultiBuffer copy = new MultiBuffer(duplicatedBuffers, chunkSize);
        copy.position = this.position;
        copy.limit = this.limit;
        return copy;
```

Reply:
Yeah, that's indeed much better. No idea why I didn't go with that. 🤦‍♀️